src/java/org/apache/cassandra/db/compaction/unified/AdaptiveController.java

```diff
@@ -0,0 +1,330 @@
+    private final int maxW;
+    private final double gain;
+    private final int minCost;
+
+    private volatile int W;
+    private volatile long lastChecked;
+
+    @VisibleForTesting
+    public AdaptiveController(TimeSource timeSource,
+                              Environment env,
+                              int W,
+                              double survivalFactor,
+                              long dataSetSizeMB,
+                              int numShards,
+                              long minSstableSizeMB,
+                              long flushSizeOverrideMB,
+                              double maxSpaceOverhead,
+                              int maxSSTablesToCompact,
+                              long expiredSSTableCheckFrequency,
+                              boolean ignoreOverlapsInExpirationCheck,
+                              int intervalSec,
+                              int minW,
+                              int maxW,
+                              double gain,
+                              int minCost)
+    {
+        super(timeSource, env, survivalFactor, dataSetSizeMB, numShards, minSstableSizeMB, flushSizeOverrideMB, maxSpaceOverhead, maxSSTablesToCompact, expiredSSTableCheckFrequency, ignoreOverlapsInExpirationCheck);
+
+        this.W = W;
+        this.intervalSec = intervalSec;
+        this.minW = minW;
+        this.maxW = maxW;
+        this.gain = gain;
+        this.minCost = minCost;
+    }
+
+    static Controller fromOptions(Environment env,
+                                  double survivalFactor,
+                                  long dataSetSizeMB,
+                                  int numShards,
+                                  long minSstableSizeMB,
+                                  long flushSizeOverrideMB,
+                                  double maxSpaceOverhead,
+                                  int maxSSTablesToCompact,
+                                  long expiredSSTableCheckFrequency,
+                                  boolean ignoreOverlapsInExpirationCheck,
+                                  Map<String, String> options)
+    {
+        int W = options.containsKey(STARTING_W) ? Integer.parseInt(options.get(STARTING_W)) : DEFAULT_STARTING_W;
+        int minW = options.containsKey(MIN_W) ? Integer.parseInt(options.get(MIN_W)) : DEFAULT_MIN_W;
+        int maxW = options.containsKey(MAX_W) ? Integer.parseInt(options.get(MAX_W)) : DEFAULT_MAX_W;
+        int intervalSec = options.containsKey(INTERVAL_SEC) ? Integer.parseInt(options.get(INTERVAL_SEC)) : DEFAULT_INTERVAL_SEC;
+        double gain = options.containsKey(GAIN) ? Double.parseDouble(options.get(GAIN)) : DEFAULT_GAIN;
+        int minCost = options.containsKey(MIN_COST) ? Integer.parseInt(options.get(MIN_COST)) : DEFAULT_MIN_COST;
+
+        if (minW >= maxW || W < minW || W > maxW)
+            throw new IllegalArgumentException(String.format("Invalid configuration for W: %d, min: %d, max: %d", W, minW, maxW));
```

Comment:
These checks are usually in `validateOptions`.

Reply:
Fixed in d5ac5f96fe64bff1c701595dd470534e7bcd1c9f.